### PR TITLE
feature/approve eip 681

### LINKED ISSUE #1828: EIP-681 URL and QRcode for "approve" function does not work
**Bug Description**

The following EIP-681 compliant URL for "**approve**" does not work:

https://metamask.app.link/send/pay-contractAddress/approve?address=receiverAddress&uint256=1e2

Please refer to the discussion on this "closed" issue (https://github.com/MetaMask/metamask-mobile/issues/1059). The QR code that is generated in accordance with EIP-681 for "approve" method causes incorrect messages to be displayed on the MetaMask Mobile v1.0.1 and the transactions eventually fail to execute.

This EIP-681 functionality was working in earlier beta releases of MetaMask in January-February 2020 time frame. I had tested it extensively. The functionality probably broke after April of 2020.

 **To Reproduce**

Use the following URL on your iPhone with MetaMask 1.0.1 installed.
https://metamask.app.link/send/pay-0x3dD3DfaAdA4d6765Ae19b8964E2BAC0139eeCb40@4/approve?0x0301dcF5Cf918a6E506AEa3C8FC6c57D8EbB8B92&uint256=50023e2

The purpose of the URL is to approve a smart contract to spend certain amount of an ERC20 token from the user's wallet.

The MetaMask-Mobile should seek user's permission to allow the smart contract permission to spend ERC20 tokens from the user's wallet. However, the MetaMask Mobile screen shows the following erroneous message and fails to process the EIP-681 request correctly.

Screenshot 2020-09-08 at 11 02 47 PM

**Next create an "approve" function at the following MetaMask Deep Link Generator:**
https://brunobar79.github.io/eip681-link-generator/# 

Scan the QRcode. It will fail to process the "approve" transaction.

screenshot2

When you scan the QRcode above using MetaMask Mobile, you get the following screen:

IMG_2201

This is incorrect and confusing. 

**Expected Behavior**

The correct message should be that "Authorize xxx to spend yyy ERC20 tokens from your wallet?" "xxx" in the previous sentence refers to the owner of the smart contract and "yyy ERC20" refers to "number of <TOKEN-NAME>". If the user approves the transaction, the smart contract is authorized to withdraw ERC20 tokens from the wallet. However, after scanning the QR code, we get the above screen (which is incorrect) and the "approve" transaction fails.

**Smart Phone**
Noticed this bug on iPhone X, iOS 13.6.1 MetaMask Mobile v1.0.1.

**Criticality**
"**approve**" is a highly critical ERC20 method required for several Dapps. Some of the common use cases include, agreeing to submit earnest money for an auction. The auctioneer will withdraw the money only if the user's bid is accepted. Other instances would include when the "approver" wants the "approvee" to withdraw the money under certain circumstances.

This issue is visible to all users and to anyone who is using EIP-681.

Please take note that in the absence of an SDK for MetaMask mobile, EIP-681 is the only way of interfacing with an app on the mobile device. Therefore, this is a highly critical issue, which needs to be resolved quickly.

### PR DESCRIPTION
**Pull request by estebanmino**
**Description**

_Write a short description of the changes included in this pull request._

**Checklist**

* [ ] There is a related GitHub issue
* [ ] Tests are included if applicable
* [ ] Any added code is fully documented

**Issue**

Resolves #1828

### CODE REVIEW

**Review by andrepimenta — APPROVED** [review]
LGTM! 👍

File: app/core/DeeplinkManager.js [hunk 1]
```diff
@@ -9,13 +9,14 @@ import AppConstants from './AppConstants';
 import Engine from './Engine';
 import { generateApproveData } from '../util/transactions';
 import { strings } from '../../locales/i18n';
+import { getNetworkTypeById } from '../util/networks';
```
Comment (nicc938) [inline app/core/DeeplinkManager.js:12]:
```suggestion
import { getNetworkTypeById } from '../util/networks';
```

File: app/core/DeeplinkManager.js [hunk 2]
```diff
@@ -34,14 +35,20 @@ class DeeplinkManager {
 			// add approve transaction
 			const {
 				parameters: { address, uint256 },
-				target_address
+				target_address,
```
Comment (nicc938) [inline app/core/DeeplinkManager.js:38]:
```suggestion
				target_address,
```

File: app/core/DeeplinkManager.js [hunk 3]
```diff
@@ -34,14 +35,20 @@ class DeeplinkManager {
...
+				chain_id
```
Comment (nicc938) [inline app/core/DeeplinkManager.js:39]:
```suggestion
				chain_id
```

**Review by ibrahimtaveras00 — APPROVED** [review]
QA Passed 👍

File: app/components/Nav/Main/index.js [hunk 4]
```diff
@@ -851,7 +851,7 @@ class Main extends PureComponent {
 			}
 
 			if (data && data.substr(0, 10) === APPROVE_FUNCTION_SIGNATURE) {
-				this.props.toggleApproveModal();
+				!this.props.approveModalVisible && this.props.toggleApproveModal();
 			} else {
```
Comment (nicc938) [inline app/components/Nav/Main/index.js:855]: #1859

**Review by nicc938 — COMMENTED** [review]
ดาวน์โหลด (1)

**Comment by nicc938** [comment]
#1847

**Pull request merged**